WIP: Adding che 6 compatibility

ScrewTSW opened the pull request:
@ibuziuk 
models are now compatible with V6, created methods to convert V6 to legacy.

DO NOT MERGE YET.

ibuziuk commented on src/main/java/io/fabric8/che/starter/model/request/WorkspaceCreateParams.java:
```diff
@@ -58,4 +61,12 @@ public void setDescription(String description) {
         this.description = description;
     }
 
+    public String getGithubToken() {
```
Could you please explain why github token was added to create workspace params ? AFAIK, fabric8-wit does not provide this info to che-starter

ScrewTSW replied:
it was originally requested from the keycloak server I think. so to add an option to circumvent it, I've made it possible to pass github token directly to che-starter.

It can definitely be removed if that's a bad decision on my end.

ibuziuk replied:
no worries, I understand why have you decided to do it, but this should be removed since we can not just break / change API between wit & che-starter

ibuziuk commented on src/main/java/io/fabric8/che/starter/controller/WorkspaceController.java:
```diff
@@ -95,10 +95,12 @@ public Workspace create(@RequestParam String masterUrl, @RequestParam String nam
         KeycloakTokenValidator.validate(keycloakToken);
-        String gitHubOAuthToken = keycloakClient.getGitHubToken(keycloakToken);
-        return createWorkspace(gitHubOAuthToken, keycloakToken, params);
+        if (params.getGithubToken()==null) {
```
spaces before and after `==`. and still do not understand why github token is coming from params

ibuziuk commented on src/main/java/io/fabric8/che/starter/client/WorkspaceClient.java:
```diff
@@ -155,12 +153,18 @@ public Workspace createWorkspace(String keycloakToken, String stackId, String re
+            workspace = gson.fromJson(workspaceResponseRaw.getBody(), Workspace.class);
+        } catch (JsonSyntaxException e) {
+            LOG.warn("Could not deserialize che server response, possibly v6");
+            WorkspaceV6 workspaceV6 = gson.fromJson(workspaceResponseRaw.getBody(), WorkspaceV6.class);
```
What about creating adding separate v6 -> v5 workspace converter class (or simply add relevant method to WorkspaceHelper / but having separate class is better IMO) which would do the following:

`public Workspace getLegacyV5WorkspaceFormat(final WorkspaceV6 workspaceV6)` 

Also, please add relevant javadoc and cover all converting functionality with JUnit tests

ScrewTSW replied:
done.

ibuziuk commented on src/main/java/io/fabric8/che/starter/controller/WorkspaceController.java:
```diff
@@ -95,10 +95,8 @@ public Workspace create(@RequestParam String masterUrl, @RequestParam String nam
             GitHubOAthTokenException, KeycloakException, WorkspaceNotFound {
-
         KeycloakTokenValidator.validate(keycloakToken);
-        String gitHubOAuthToken = keycloakClient.getGitHubToken(keycloakToken);
```
for debugging purposes please leave the var extruction

ScrewTSW replied:
I'm not sure what do you mean by that.

ibuziuk replied:
` String gitHubOAuthToken = keycloakClient.getGitHubToken(keycloakToken);`
please leave it like that

ScrewTSW replied:
Renamed to remain consistent with the rest of the file.

ScrewTSW replied:
renamed for consistency in file.

ibuziuk commented on src/main/java/io/fabric8/che/starter/util/WorkspaceLegacyFormatAdapter.java:
```diff
@@ -0,0 +1,83 @@
+    private static final String SNAPSHOT = "/snapshot";
+
+    public static Workspace convertV6ToLegacyFormat(final WorkspaceV6 workspaceV6) {
+        Workspace response = new Workspace();
```
response does no sound like a good name for var

ibuziuk commented on src/main/java/io/fabric8/che/starter/util/WorkspaceLegacyFormatAdapter.java:
```diff
@@ -0,0 +1,83 @@
+        return response;
+    }
+
+    private static List<WorkspaceLink> convertLinksToLegacy(WorkspaceV6 workspaceV6) {
```
lgtm

ScrewTSW commented on src/main/java/io/fabric8/che/starter/model/request/WorkspaceCreateParams.java:
```diff
@@ -13,13 +13,15 @@
  *
  */
 @JsonIgnoreProperties(ignoreUnknown = true)
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
It's not necessary, I put it there in case, that if something does not match up, it wont break the rest endpoint.

ibuziuk replied:
This annotation should be removed since params like branch & description are optional. Changes for this class should be reverted

ibuziuk commented:
[test]

ibuziuk commented:
@ScrewTSW PR build is currently failing - see code conventions & license in Readme https://github.com/redhat-developer/che-starter#code-conventions

ibuziuk commented on src/main/java/io/fabric8/che/starter/util/WorkspaceComparator.java:
```diff
@@ -0,0 +1,147 @@
+import java.util.Comparator;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class WorkspaceComparator implements Comparator<Workspace> {
```
Shouldn't this utility be in `src/ test` ?

ScrewTSW replied:
if it will never be used in che-starter itself, I think I should move it there.

ibuziuk commented on src/test/java/io/fabric8/che/starter/util/WorkspaceLegacyFormatAdapterTest.java:
```diff
@@ -0,0 +1,229 @@
+
+        WorkspaceEnvironments v5CorrectFullConfigEnvironments = new WorkspaceEnvironments();
+        v5CorrectFullConfigEnvironments.setDefaultEnv(v5CorrectFullConfigEnvironmentDefault);
+        /*========================*/
```
do we really need this comments ? imo it just pollutes the test

ibuziuk commented on src/test/java/io/fabric8/che/starter/util/WorkspaceLegacyFormatAdapterTest.java:
```diff
@@ -0,0 +1,229 @@
+
+        /*====================*
+         * WORKSPACE PROJECTS *
+         *====================*/
```
this type of comment look ok

ibuziuk commented:
@ScrewTSW build is back to normal. Are you planning any other changes to this PRv ?

ScrewTSW commented:
@ibuziuk yes, i still have two tests to test compatibility in and then final test with che-starter against both che5 and che6 before we merge this. I really don't want to rush a crucial change.

ibuziuk commented on src/main/java/io/fabric8/che/starter/client/WorkspaceClient.java:
```diff
@@ -109,11 +113,21 @@ public void waitUntilWorkspaceIsStopped(Workspace workspace, String keycloakToke
+        ResponseEntity<String> responseRaw = template.exchange(url, HttpMethod.GET, null, String.class);
+        List<Workspace> response = new ArrayList<>();
+        try {
+            List<Workspace> tmp = gson.fromJson(responseRaw.getBody(), new TypeToken<ArrayList<Workspace>>(){}.getType());
```
tmp does not sound like a good name for var - consider `workspaces`

ScrewTSW replied:
done.

ibuziuk commented on src/main/java/io/fabric8/che/starter/client/WorkspaceClient.java:
```diff
@@ -109,11 +113,21 @@ public void waitUntilWorkspaceIsStopped(Workspace workspace, String keycloakToke
+            tmp.parallelStream().forEach(response::add);
+        } catch (Exception e) {
+            LOG.info("Unable to deserialize workspace list, probably V6 format.");
+            List<WorkspaceV6> tmp = gson.fromJson(responseRaw.getBody(), new TypeToken<ArrayList<WorkspaceV6>>(){}.getType());
```
`tmp` does not sound like a good name

ScrewTSW replied:
done.

ibuziuk commented:
@ScrewTSW wondering if `getWorkspaceById / getWorkspaceByName` are going to work correctly against che6

ibuziuk commented on src/test/java/io/fabric8/che/starter/util/WorkspaceLegacyFormatAdapterTest.java:
```diff
@@ -0,0 +1,225 @@
+            InputStream workspaceV6JsonStream = getClass().getClassLoader().getResourceAsStream("workspacev6.json");
+            workspaceV5Json = IOUtils.toString(workspaceV5JsonStream, StandardCharsets.UTF_8);
+            workspaceV6Json = IOUtils.toString(workspaceV6JsonStream, StandardCharsets.UTF_8);
+        } catch (IOException e) {
```
why should we catch exceptions in tests ?

ScrewTSW replied:
That's in case that the source json files cannot be loaded because of IO errors.
I log an error so that we are not pointlessly flooded with stack and exit with status `-1`

ScrewTSW replied:
can remove

ScrewTSW replied:
done.

ibuziuk commented on src/test/java/io/fabric8/che/starter/util/WorkspaceLegacyFormatAdapterTest.java:
```diff
@@ -0,0 +1,225 @@
+            workspaceV6Json = IOUtils.toString(workspaceV6JsonStream, StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            LOG.error("Could not read test input JSON files.");
+            System.exit(-1);
```
This is not required

ScrewTSW replied:
had to add `throws` signature because of IOUtils

ibuziuk commented on src/main/java/io/fabric8/che/starter/model/workspace/WorkspaceConfigV6.java:
```diff
@@ -0,0 +1,96 @@
+    private List<WorkspaceLink> links;
+    private List<Project> projects;
+
+    public WorkspaceConfigV6() {
```
Why do we need constructor for DTO ?

ScrewTSW replied:
copied from V5 ;)

ScrewTSW replied:
will clean up both

ScrewTSW replied:
done.

ScrewTSW commented on src/main/java/io/fabric8/che/starter/model/workspace/WorkspaceEnvironmentV6.java:
```diff
@@ -0,0 +1,46 @@
+    private Map<String, WorkspaceMachineV6> machines;
+
+    public WorkspaceEnvironmentV6() {
+        setMachines(new HashMap<String, WorkspaceMachineV6>());
```
also copied from V5

ScrewTSW replied:
will clean up both

ScrewTSW replied:
done.

ibuziuk commented:
[test]

ScrewTSW commented:
```
@ScrewTSW wondering if getWorkspaceById / getWorkspaceByName are going to work correctly against che6
```
Names and IDs behave the same. It should work.

ibuziuk commented:
just got NPE when running your PR locally:

```
Tests in error: 
  WorkspaceLegacyFormatAdapterTest.getWorkspaceLegacyFormat:226 » NullPointer

```

ibuziuk commented:
oh, I'm afraid it is even worse:
Tests in error: 
  

> WorkspaceClientTest.createAndDeleteWorkspace:73->lambda$createAndDeleteWorkspace$2:73 NullPointer
>   WorkspaceLegacyFormatAdapterTest.getWorkspaceLegacyFormat:226 » NullPointer

This is what I got on the second run

ibuziuk commented on src/main/java/io/fabric8/che/starter/client/WorkspaceClient.java:
```diff
@@ -109,11 +113,21 @@ public void waitUntilWorkspaceIsStopped(Workspace workspace, String keycloakToke
+        List<Workspace> response = new ArrayList<>();
+        try {
+            List<Workspace> workspaces = gson.fromJson(responseRaw.getBody(), new TypeToken<ArrayList<Workspace>>(){}.getType());
+            workspaces.parallelStream().forEach(response::add);
```
parallelStream should be avoided in multi-thread environment.

ScrewTSW replied:
done.

ibuziuk commented:
[test]

ibuziuk commented:
[test]

ibuziuk commented:
@ScrewTSW when running your PR locally against che6 I got the following test failures:
 
```
Results :

Tests in error: 
  WorkspaceClientTest.getWorkspaceByIdAndDelete:85 » HttpMessageNotReadable Coul...
  WorkspaceClientTest.getWorkspaceByNameAndDelete:102 » HttpMessageNotReadable C...
  WorkspacePreferencesClientTest.setCommitterInfo:55 » HttpClientError 405 null
  GitHubTokenClientTest.setGitHubToken:57 » GitHubOAthToken Error setting GitHub...
```

WorkspacePreferencesClientTest / GitHubTokenClientTest failures are expected but it is not expected to have failures in WorkspaceClientTest. Could you please take a look ?

ScrewTSW commented:
@ibuziuk sure, will do.

ibuziuk commented on src/main/java/io/fabric8/che/starter/model/workspace/WorkspaceV6.java:
```diff
@@ -0,0 +1,70 @@
+    // Runtime is returned only on get workspace by composite key, in collection call, it is not
+    private WorkspaceRuntime runtime;
+    private WorkspaceConfigV6 config;
+    //TODO: config -> agents is now called installers
```
is this TODO relevant ?

ScrewTSW replied:
sorry, forgot to remove that one.

ibuziuk commented on src/main/java/io/fabric8/che/starter/model/workspace/WorkspaceV6.java:
```diff
@@ -0,0 +1,70 @@
+
+    private String id;
+    private String status;
+    // Runtime is returned only on get workspace by composite key, in collection call, it is not
```
the comment is not clear for me

ScrewTSW replied:
I have no idea what it means either -> original DTO had it

ibuziuk commented on src/test/java/io/fabric8/che/starter/util/WorkspaceLegacyFormatAdapterTest.java:
```diff
@@ -0,0 +1,238 @@
+        v5CorrectFull.setLinks(v5CorrectFullLinks);
+    }
+
+    @Repeat(value = 1000000)
```
not sure we need repeat for this particular test, since there should not be race conditions during deserialization / comparison - if there is one, there is a problem in design

ScrewTSW replied:
ok.

ibuziuk commented:
also please rebase to the latest master

ibuziuk commented:
PR is almost ready to be merged. Please, describe how was it tested against `che6` ?

ScrewTSW commented:
rebased to the latest master, hopefully it will help with the client test, because I have no idea what went wrong there.

Also when I was developing it I was testing it against minishift with che-add-on che 6.2.0-SNAPSHOT
Then later tested through the .properties file where I changed it to a che6 server.

ibuziuk commented:
Acceptance criteria:
- all tests are passing against che5
- all tests are passing against che6 

- Manual testing of starting new  (POST) / and existing (PATCH) workpspace - this is not covered by integration tests against che5 / che6

ScrewTSW commented:
@ibuziuk I could write a test case for the starting new and starting existing.

ScrewTSW commented:
[test]

ibuziuk commented:
[test]

ibuziuk commented:
[test]

ibuziuk commented:
[test]

ibuziuk commented:
[test]

ibuziuk commented:
[test]
